src/economy/shipping_schedule.cc

```diff
@@ -0,0 +1,1704 @@
+/*
+ * Copyright (C) 2019-2020 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+
+#include "economy/shipping_schedule.h"
+
+#include <memory>
+#include <set>
+
+#include "economy/expedition_bootstrap.h"
+#include "economy/portdock.h"
+#include "economy/ship_fleet.h"
+#include "io/fileread.h"
+#include "io/filewrite.h"
+#include "logic/game.h"
+#include "logic/game_data_error.h"
+#include "logic/map_objects/checkstep.h"
+#include "logic/map_objects/tribes/ship.h"
+#include "logic/player.h"
+#include "map_io/map_object_loader.h"
+#include "map_io/map_object_saver.h"
+
+namespace Widelands {
+
+/*******************************************************************************
+              Weighting factors to tweak the algorithm's decisions
+*******************************************************************************/
+
+// Performance tradeoff in the First Pass
+constexpr uint32_t kActualDurationsRecalculationInterval = 60 * 1000;
+
+// Give a ship the highest score for assigning wares to it if it will
+// arrive at the port of interest within the next 10 s,
+// or the lowest score if this will take longer than 10 min
+constexpr Duration kWonderfullyShortDuration = 10 * 1000;   // 10 s
+constexpr Duration kHorriblyLongDuration = 10 * 60 * 1000;  // 10 min
+
+// Only assign wares to a ship in 5.1 if it's score is higher than a certain
+// threshold based on minimal distance. Ships with lower scores will only be
+// accepted if not enough idle ships are found.
+constexpr uint16_t kMinScoreForImmediateAcceptFactor = 14;  // NOCOM needs testing
```

Comment:
Not really. If players observe that the schedule tends to assign many idle ships to a route even if ships are already coming then this can be tweaked by increasing this number. If players observe that the schedule waits for ages for a ship even when there are idle ships available this can be tweaked by decreasing this number.
Since nobody complained in testing I assume 14 is a fairly good magic number here.